### PR TITLE
Use local checkpoint to calculate min translog gen for recovery

### LINKED ISSUE #49970: Too many translog files open, and reach ulimit setting(65535)
**Describe the feature**:
Hi, 
I found 'too many files open' problem in my ES enviroment.  And I use the commad `ls -al /proc/ES_PID/fd` , found that ES opened particularly large number of translog files.  
After some search on this problem , I tried to restart  a runing normally ES (6.6.1).  I found that ，**every time ES is restarted, there will be one more translog file on some shards**. And all the translogs have no operation data, like  below( The server time is incorrect，but I don't think it's concern ):

image

**Elasticsearch version** (`bin/elasticsearch --version`):
Version: 6.6.1, Build: default/tar/1fd8f69/2019-02-13T17:10:04.160291Z, JVM: 1.8.0_60
**Plugins installed**: []
No Plugins
**JVM version** (`java -version`):
java version "1.8.0_60"
Java(TM) SE Runtime Environment (build 1.8.0_60-b27)
**OS version** (`uname -a` if on a Unix-like system):
Centos 7.6
Linux 3.10.0-957.5.1.el7.x86_64 #1 SMP Fri Feb 1 14:54:57 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
**Description of the problem including expected versus actual behavior**:
Empty translog files(no operation data) keep increasing.
Old translog exist beyond 12 hours(index.translog.retention.age = 12h).
I think empty translog files should be deleted or don't creat new translog files.
**Steps to reproduce**:
restart ES node
Please include a *minimal* but *complete* recreation of the problem, including
(e.g.) index creation, mappings, settings, query etc.  The easier you make for
us to reproduce it, the more likely that somebody will take the time to look at it.

 1.  defautl index setting as follow:
`"settings":{  
              "index.number_of_replicas":"0",
              "index.refresh_interval":"10s",
              "index.number_of_shards":"1",
              "index.routing.allocation.require.tag":"hot",
              "index.indexing.slowlog.level": "info",
              "index.indexing.slowlog.threshold.index.trace": "500ms",
              "index.indexing.slowlog.threshold.index.debug": "2s",
              "index.indexing.slowlog.threshold.index.info": "5s",
              "index.indexing.slowlog.threshold.index.warn": "10s",
              "index.search.slowlog.level": "info",
              "index.search.slowlog.threshold.fetch.trace": "200ms",
              "index.search.slowlog.threshold.fetch.debug": "500ms",
              "index.search.slowlog.threshold.fetch.info": "800ms",
              "index.search.slowlog.threshold.fetch.warn": "1s",
              "index.search.slowlog.threshold.query.trace": "500ms",
              "index.search.slowlog.threshold.query.debug": "2s",
              "index.search.slowlog.threshold.query.info": "5s",
              "index.search.slowlog.threshold.query.warn": "10s",
               "index.translog.durability": "async",
               "index.translog.flush_threshold_size": "5000mb",
               "index.translog.sync_interval": "120s",
               "index.mapping.ignore_malformed": true
   }
`
 2.  Auto create index use the dynamic mapping, and index name is created by date, only the index created today is active. Indices  created before today have no data to index.
 3.  Restart ES, the translog of indices  created before today (no data to index) keep increasing

If I missed some configuration， please let me know.
I'd be happy to provide additional details, whatever is needed.

Thanks!

### PR DESCRIPTION
**Pull request by dnhatn**
Today we use the translog_generation of the safe commit as the minimum required translog generation for recovery. This approach has a limitation, where we won't be able to clean up translog unless we flush. Reopening an already recovered engine will create a new empty translog, and we leave it there until we force flush.

This commit removes the translog_generation commit tag and uses the local checkpoint of the safe commit to calculate the minimum required translog generation for recovery instead.

Closes #49970

### CODE REVIEW

**Review by ywelsch — COMMENTED** [review]
I think that the `uncommittedSizeInBytes` and `uncommittedOperations` metrics are pretty useless today, as they are not a measure of how much data needs to be recovered/replayed after a crash. I would rather base these metrics on the local checkpoint of the safe commit in all cases, and completely disregard the last commit (which is irrelevant for recovery).

**Comment by dnhatn** [comment]
> I think that the uncommittedSizeInBytes and uncommittedOperations metrics are pretty useless today, as they are not a measure of how much data needs to be recovered/replayed after a crash. I would rather base these metrics on the local checkpoint of the safe commit in all cases, and completely disregard the last commit (which is irrelevant for recovery).

++. I pushed https://github.com/elastic/elasticsearch/pull/51905/commits/c4bcceafc01ddbb745063fe0cfbea246079c5e64.

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 1]
```diff
@@ -458,23 +458,22 @@ public void skipTranslogRecovery() {
...
+            }
+        } else {
+            opsRecovered = 0;
         }
         // flush if we recovered something or if we have references to older translogs
         // note: if opsRecovered == 0 and we have older translogs it means they are corrupted or 0 length.
         assert pendingTranslogRecovery.get() : "translogRecovery is not pending but should be";
         pendingTranslogRecovery.set(false); // we are good - now we can commit
         if (opsRecovered > 0) {
-            logger.trace("flushing post recovery from translog. ops recovered [{}]. committed translog id [{}]. current id [{}]",
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: restored in https://github.com/elastic/elasticsearch/pull/51905/commits/2c0b2573dc8ee7aa21c418b0319dad3e276f1c8a

**Comment by ywelsch** [comment]
There are relevant test failures here

File: server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java [hunk 2]
```diff
@@ -5664,8 +5647,8 @@ public void testRecoverFromLocalTranslog() throws Exception {
                 for (Engine.Operation op : operations) {
                     applyOperation(engine, op);
                     if (randomBoolean()) {
+                        globalCheckpoint.set(randomLongBetween(globalCheckpoint.get(), engine.getProcessedLocalCheckpoint()));
```
Comment (henningandersen) [inline server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java]: I wonder why this change is necessary? I find it a bit counter intuitive to update that before sync'ing translog, but I am sure there is a good reason.

**Comment by dnhatn** [comment]
@ywelsch Thanks for looking at the test failures. I've addressed them in https://github.com/elastic/elasticsearch/pull/51905/commits/5ea2cae112c15f1efd2ff13ddd22e2469e9762a3.

File: server/src/test/java/org/elasticsearch/index/shard/IndexShardIT.java [hunk 3]
```diff
@@ -416,7 +416,7 @@ public void testStressMaybeFlushOrRollTranslogGeneration() throws Exception {
         final Settings settings;
         if (flush) {
             // size of the operation plus two generations of overhead.
```
Comment (henningandersen) [inline server/src/test/java/org/elasticsearch/index/shard/IndexShardIT.java]: I am guessing that the change in the next line is due to only keeping one generation now? In that case, this comment needs updating I think.

File: server/src/test/java/org/elasticsearch/index/engine/NoOpEngineTests.java [hunk 4]
```diff
@@ -168,49 +166,28 @@ public void testTrimUnreferencedTranslogFiles() throws Exception {
...
-            Map<String, String> lastCommittedUserData = indexCommitRef.getIndexCommit().getUserData();
-            lastCommitedTranslogGeneration = Long.parseLong(lastCommittedUserData.get(Translog.TRANSLOG_GENERATION_KEY));
-            assertThat(lastCommitedTranslogGeneration, equalTo(currentFileGeneration));
-        }
-
-        assertThat(Translog.readMinTranslogGeneration(translogPath, translogUuid), equalTo(minFileGeneration));
-        assertThat(noOpEngine.getTranslogStats().estimatedNumberOfOperations(), equalTo(softDeleteEnabled ? 0 : numDocs));
-        assertThat(noOpEngine.getTranslogStats().getUncommittedOperations(), equalTo(0));
-
         noOpEngine.trimUnreferencedTranslogFiles();
```
Comment (henningandersen) [inline server/src/test/java/org/elasticsearch/index/engine/NoOpEngineTests.java]: Should we check prior to this that we have some operations to trim, except in the case where the last thing done was a flush?

**Review by henningandersen — APPROVED** [review]
This LGTM. I am not familiar enough with all 6.x development that I can confidently say that this will work in both rolling and full restart upgrades, but my search for an issue did not reveal anything. So better wait for Yannick's approval too.

File: server/src/main/java/org/elasticsearch/index/engine/NoOpEngine.java [hunk 5]
```diff
@@ -137,31 +138,26 @@ public void trimUnreferencedTranslogFiles() {
         try (ReleasableLock lock = readLock.acquire()) {
             ensureOpen();
             final List<IndexCommit> commits = DirectoryReader.listCommits(store.directory());
-            if (commits.size() == 1) {
+            if (commits.size() == 1 &&
+                translogStats.getUncommittedOperations() == 0 &&
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/engine/NoOpEngine.java]: Good catch. That is not needed. I removed it in https://github.com/elastic/elasticsearch/pull/51905/commits/63c25e2adc5802bd1bc128e3204b844e6a83a7bf.

File: server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java (hunk 2 above)
Comment (dnhatn) [reply server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java]: I've reverted this in https://github.com/elastic/elasticsearch/pull/51905/commits/630da78ad629bf1a9fb7af78c9dcee77cb7c2b66.

File: server/src/test/java/org/elasticsearch/index/engine/NoOpEngineTests.java (hunk 4 above)
Comment (dnhatn) [reply server/src/test/java/org/elasticsearch/index/engine/NoOpEngineTests.java]: ++. Added in https://github.com/elastic/elasticsearch/pull/51905/commits/02103b77d611ab3fc10d0d90299f28c7508b9cd8.

File: server/src/test/java/org/elasticsearch/index/shard/IndexShardIT.java (hunk 3 above)
Comment (dnhatn) [reply server/src/test/java/org/elasticsearch/index/shard/IndexShardIT.java]: fixed in https://github.com/elastic/elasticsearch/pull/51905/commits/8c91dcfe43d55bef8c109885527cc7a3584f74a4.

**Review by ywelsch — APPROVED** [review]
LGTM

**Comment by dnhatn** [comment]
@ywelsch @henningandersen Thanks for reviewing.

**Pull request merged**